### ISSUE TITLE
Reverse proxy black screen with 2.3.x on JF 10.7.x

### ISSUE THREAD

**Issue by Nazar78** [jellyfin/jellyfin-android#494]
### Describe the bug

Hi all, 

Referring to the reverse proxy for "Nginx from a subdomain" as stated in the documentation https://jellyfin.org/docs/general/networking/nginx.html, 2.3.0 on JF 10.7.6 can only launched the first time after login. Close the app then reopen will result in a black screen permanently unless reinstalled or data cleared.

Things that has been tested:

1. Reverse proxy works correctly on previous version before 2.3.0.
2. Reverse proxy works correctly from webUI and Tizen with no errors in developer tools.
3. Tested with the exact snippet copied from the JF reverse proxy documentation except for these changes, "set $jellyfin 127.0.0.1;" and commented out the resolver.
4. Reverse proxy tested on both port 80 and 443 (Let's Encrypt).
5. Reverse proxy tested on Note 20 Ultra Android 11 and Android Studio API 29/30 clean image.
6. I'm not attaching any logcat or debug logs because there's no specific errors comparing to connecting directly to port 8096 which works correctly. Just a black screen when using the reverse proxy. I could see the websocket activities though in the logcat, server and Android Studio debug log that's just it. Hope someone could help to replicate the reverse proxy and point out where is the issue.
7. Last edit: Not sure if it matters but the subdomain is on IPV6 fallback to IPV4.

Thank you.

All running on debian 11 (bullseye)
JF 10.7.6
Nginx 1.18.0

### Application version

2.3.0

### Where did you install the app from?

Google Store

### Device information

Samsung Note 20 Ultra

### Android version

Android 11

### Jellyfin server version

10.7.6

### Which video player implementations does this bug apply to?

- [ ] Web player (default)
- [ ] Integrated player (ExoPlayer)
- [ ] External player (VLC, mpv, MX Player)

**Comment by Nazar78** [comment]
Received the version 2.3.2 from play store update today. Cleared the data and login was successful as expected via nginx reverse proxy. But upon reopening the app again, I still got the same blank screen issue unless I clear the data. With direct port, it works as expected.

What could be the reason it doesn't load after closing? I can't find anything in the debug. Any pointers appreciated.

**Comment by Maxr1998** [comment]
What system do you host your server on? Linux or Windows? And what did you enter into the app? IP? Hostname? Including protocol or port or without?

**Comment by Maxr1998** [comment]
It might be that it persists the wrong hostname, and thus isn't able to connect upon restarting the app. I have no idea why it would do that though..

**Comment by Nazar78** [comment]
> I removed the logs from your comment since they accidentally leaked both your server domain and access tokens, I'd advise you to log out all sessions on your Jellyfin server to be safe.

Ok thank you.

> I wonder whether the WB_REVISION query string could cause the load to fail.

I'm not sure about this. From my proxy request logs, it seems to load with proper HTTP 20x/30x. Any way for me to help check? From my further testing I noticed the followings (please note that these are tested on a fresh clean JF/Nginx server on Ubuntu 20.04.3):

1. The app has no black screen issue with reverse proxy in `HTTP` but happens consistently with reverse proxy in `HTTPS`. So there's seems to be issues in reverse proxy with HTTPS. This is not a self-signed but Let's Encrypt wildcard cert. `HTTPS` works correctly in web browsers, Tizen/Android TV and the older JF app before 2.3.x. I'll try to dig in further.
2. A minor thing, the app doesn't handle this redirection correctly. In the documentation: https://jellyfin.org/docs/general/networking/nginx.html on this particular redirect block, it works on the web but not on the app. `    location = / {
        return 302 https://$host/web/;
    }` Also the `https` should be replaced with `$schema` to detect http or https request.

**Comment by niekcandaele** [comment]
Hey 

Seeing a similar issue.

Had Jellyfin setup with a reverse proxy (https://jellyfin.example.com). Accessing via the browser works. We had it set up via the app too, which worked for a long time. However, at some point I assume the app auto updated and now everyone just gets a black screen.

I've removed the local app data from my phone so I get the original connection flow again.

I've tried every way I can think of to connect. Even opening up the raw JF ports, bypassing our reverse proxy (and TLS unfortunately)

- `https://jellyfin.example.com`
- `http://jellyfin.example.com`
- `https://jellyfin.example.com:8096`
- `http://jellyfin.example.com:8096`
- `http://<IP>:8096`

None of these worked.

The DNS is handled by Cloudflare, but without CF features enabled (so just 'raw' DNS). It's a wildcard record.


I am 100% confident the connection set up on the server is fine. It worked for a long time, then suddenly broke :(

**Comment by Maxr1998** [comment]
Could you try to reproduce it on the debug release and take logs for it please?

**Comment by Nazar78** [comment]
Hi @Maxr1998,

I just tested the commit from master today and this time after app restarts there's no black screen, there's the circle loading for few secs then it goes back to the typical "Connect to Server" URL input view with indication in red "Connection cannot be established...". However I can never connect back to that particular JF URL again. From that same moment, I can connect and login to the local LAN JF URL, but when I tried to switch to the reverse proxy domain again, same thing occurred unless I clear the app's data.

Attached is the debug log you've requested: [org.jellyfin.mobile.debug.log](https://github.com/jellyfin/jellyfin-android/files/7079839/org.jellyfin.mobile.debug.log). This is the log after I killed the emulator API 29 JF app, then start the debug process in Android Studio.

To expedite the debugging further, please use this reverse proxy instance. It's the same JF server as in the debug log. Both JF and Nginx running in chroot on Debian 11 (bullseye).

URL: https://demo.teanazar.com
User: demo
Pass: blank, no pass required.

**Comment by Maxr1998** [comment]
Thanks a lot for the additional logs and setting up a test instance.

I only tried the demo instance on my main device so far, but on there, everything works perfectly fine. It always connects instantly on every restart of the app.
It's weird how I'm not facing this issue, there must be something network related mixed into this bug..

**Comment by Nazar78** [comment]
Hi @Maxr1998,

Thanks again for looking into the issue. Now that you mentioned your main device, I tested mine too with the demo instance and JF app from play store 2.3.2.

I'm using dual stack IPv6/IPv4 at home and the issue is always consistent perhaps not sure if I was connected to a IPv6. So to simulate a typical IPv4 connection, I tethered my main mobile to a wifi hotspot from a second mobile. Both mobile are actually on the same telco network only different IPv4 assigned.

So in this test I got few good start up but once I switch network i.e. mobile to wifi (tethered), after few times it's black screen again permanently unless data is cleared. Please refer to the screen recording attached. 

https://user-images.githubusercontent.com/2883688/131476069-82e00d91-9eb2-4d5c-b7bd-4bd2fd0f85b7.mp4

I also tested the possibility of using plain DNS vs DoH, nothing changes so I'm out of ideas what else to test see if I can spot any irregularities. If there's anything I can help to test please do let me know.

**Comment by Maxr1998** [comment]
Did some more investigation today, and found a lead to the issue. Unfortunately, I have no idea how to actually fix it.

The app relies on a `.../main.*.bundle.js` to be loaded when starting the app, which should trigger the callback `shouldInterceptRequest` of the `WebViewClient`, from where the app can inject its own scripts. For some reason, that callback isn't executed when the bundle file was cached and is served from memory. Because it theoretically always is intercepted, it shouldn't ever get cached, but apparently specific conditions can cause the WebView to do just that and from there on break the injection. The "black screen" only happens because the WebView is invisible on start (to prevent flashes) and is set to visible when the callback is triggered.

tl;dr it's a caching/API issue in WebView that I don't know how to fix. Switching to a different entrypoint to inject the scripts (as we did before 2.3) isn't really possible since that had its own set of issues. Only possibility I see right now would be migrating to something like GeckoView which should give us more control over the webapp. I have no idea how much work that is, however..

**Comment by Nazar78** [comment]
Thanks @Maxr1998,

I also tried from the proxy's end it doesn't work either (I've reverted back to proxy caching after testing).
```
location ~* .+/main\..+\.bundle\.js {
	add_header Pragma no-cache;
	add_header Cache-Control "no-store, no-cache, must-revalidate, post-check=0, pre-check=0, max-age=0";
	add_header Expires "Thu, 01 Jan 1970 00:00:01 GMT";
	proxy_pass http://jellyfin;
	proxy_buffering off;
}
```

```
HTTP/2 200 
date: Fri, 03 Sep 2021 10:07:20 GMT
content-type: application/javascript; charset=utf-8
content-length: 997389
server: Kestrel
last-modified: Fri, 21 May 2021 02:06:52 GMT
accept-ranges: bytes
etag: "1d74de5f72abe0d"
x-response-time-ms: 0
pragma: no-cache
cache-control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0, max-age=0
expires: Thu, 01 Jan 1970 00:00:01 GMT
```
Actually just `Cache-Control: no-store` should suffice telling the WebView not to cache, but still all these still resulted in black screen eventually. Would you able to help point out where this injection took place?

**Comment by Nazar78** [comment]
Hi @Maxr1998 ,

Today I just upgraded to JF 10.7.7 and the issue still persist. 

However before upgrading, I did managed to somewhat figure out the workaround after a while messing with the codes. This also works in the previous JF 10.7.6. Seems all the injection works. The main issue as you've pointed out, was because of the caching and the injection doesn't trigger. But this seems to be specifically a service worker caching issue. Once the SW cached the assets on first successful login, there's no way to login again unless the app data is cleared because the injection doesn't trigger. 

So what I did was to destroy the SW. I managed to successfully use either of these two workarounds and would like to seek advice if they are acceptable, especially the no. 2 for PR. Do we really need the SW for the webview? Any other usage besides cache? I noticed most of the assets can load from the disk cache after destroying the SW cache. Further more SW cache won't work for http only https.

1. For the play store version on my main samsung phone, I forced my proxy server to return blank on the serviceworker.js. Then I just have to clear the app's data once and it's persistent.
```
set $androidWebviewServiceWorker 0;
if ($http_user_agent ~ .*Android.+wv\).*) {
	set $androidWebviewServiceWorker 1;
}

if ($request_uri ~ /web/serviceworker.js.*) {
	set $androidWebviewServiceWorker ${androidWebviewServiceWorker}1;
}

if ($androidWebviewServiceWorker = 11) {
	return 200;
}
```
> HTTP/2 200
> server: nginx/1.18.0
> date: Mon, 06 Sep 2021 09:40:23 GMT
> content-type: application/javascript; charset=utf-8
> content-length: 0
2. For the master version, I made another injection but it's for destroying SW. Then just launch the app once, relaunch again no need to clear the data if its being cached, although it's better to clear the data which can save a few MBs.
```
// org/jellyfin/mobile/fragment/WebViewFragment.kt: WebView.initialize()
...
            showOutdatedWebViewDialog(this)
            return
        }

        // Workaround for serviceWorker caching
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val swController = ServiceWorkerController.getInstance()
            swController.setServiceWorkerClient(object : ServiceWorkerClient() {
                override fun shouldInterceptRequest(request: WebResourceRequest): WebResourceResponse? {
                    val url = request.url
                    val path = url.path?.lowercase(Locale.ROOT) ?: return null
                    return when {
                        path.endsWith("serviceworker.js") -> {
                            WebResourceResponse("text/javascript", "utf-8",
                                ByteArrayInputStream(
                                    "// Workaround for serviceWorker caching"
                                        .toByteArray()
                                )
                            )
                        }
                        else -> null
                    }
                }
            })
        }

	webViewClient = object : WebViewClientCompat() {
...
```

Please help to review and advise, thank you.

**Comment by Maxr1998** [comment]
Thanks for further investigating this issue! Blocking the service worker seems to be a valid workaround. The proposed patch looks good, please open a PR so that I can properly review it.
(Btw, there should be a `byteInputStream()` method on String that you could use).

**Comment by Nazar78** [comment]
byteInputStream, got it and I've created the PR. Thank you!
